mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,819 @@
+            self.from_block_height,
+        )
+        .await
+        {
+            Ok(block) => block,
+            Err(stopped) => {
+                if let StopReason::RpcClient(e) = &stopped.stop_reason {
+                    log::error!("Received tx history process turned into cooldown mode due to rpc error: {e}");
+                    return Self::change_state(OnIoErrorCooldown::new(self.address.clone(), self.from_block_height));
+                }
+
+                return Self::change_state(stopped);
+            },
+        };
+
+        let last_fetched_block = cmp::max(highest_send_tx_height, highest_recieved_tx_height);
+
+        log::info!(
+            "Tx history fetching finished for {}. Last fetched block {}",
+            ctx.coin.platform_ticker(),
+            last_fetched_block
+        );
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Finished);
+        Self::change_state(WaitForHistoryUpdateTrigger::new(
+            self.address.clone(),
+            last_fetched_block,
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for TendermintInit<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        const INITIAL_SEARCH_HEIGHT: u64 = 0;
+
+        ctx.coin.set_history_sync_state(HistorySyncState::NotStarted);
+
+        if let Err(e) = ctx.storage.init(&ctx.coin.history_wallet_id()).await {
+            return Self::change_state(Stopped::storage_error(e));
+        }
+
+        let search_from = match ctx
+            .storage
+            .get_highest_block_height(&ctx.coin.history_wallet_id())
+            .await
+        {
+            Ok(Some(height)) if height > 0 => height as u64 - 1,
+            _ => INITIAL_SEARCH_HEIGHT,
+        };
+
+        Self::change_state(FetchingTransactionsData::new(
+            ctx.coin.my_address().expect("my_address can't fail"),
+            search_from,
+        ))
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> LastState for Stopped<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result {
+        log::info!(
+            "Stopping tx history fetching for {}. Reason: {:?}",
+            ctx.coin.ticker(),
+            self.stop_reason
+        );
+
+        let new_state_json = json!({
+            "message": format!("{:?}", self.stop_reason),
+        });
+
+        ctx.coin.set_history_sync_state(HistorySyncState::Error(new_state_json));
+    }
+}
+
+#[async_trait]
+impl TendermintTxHistoryOps for TendermintCoin {
+    async fn get_rpc_client(&self) -> MmResult<HttpClient, TendermintCoinRpcError> { self.rpc_client().await }
```

Comment:
@shamardy I feel its kinda related with my refactoring `rpc_clients in tendermint` task.
So we usually keep vector of clients and we have to iterate it every time to send request ([this issue](https://github.com/KomodoPlatform/atomicDEX-API/issues/1480)).
We have such problems like [here in web3](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth/web3_transport.rs#L243), in [z_rpc](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/z_coin/z_rpc.rs#L756), in [tendermint_coin](https://github.com/KomodoPlatform/atomicDEX-API/blob/iris-swap-opt/mm2src/coins/tendermint/tendermint_coin.rs#L422).
So what I want to say, may be we should have a separate module for `trait` with `Ops` to get client or iterate over urls, if client doesn't work.
Every time, when we need to impl this trait, we can create a special `structure` that contains two fields: one active client and vec of urls. So we can use this structure as field for `TendermintCoinImpl` (for example) or use it instead of `Vec<CompactTxStreamerClient<Channel>>` from `z_rpc`.
There is a bit another way. Every time we can create a special `trait` for every coin with `structure` that implements it.
@ozkanonur @sergeyboyko0791 what do you think?

Reply:
To be honest, I didn't quite understand what you mean, so will try to explain my opinion:
Regarding https://github.com/KomodoPlatform/atomicDEX-API/issues/1480, I think it could be solved within a Rpc client structure, not outside of it:
```rust
struct Web3Transport {
  // An identifier of the URI
  active_rpc_id: usize,
  id: Arc<AtomicUsize>,
  uris: Vec<http::Uri>,
  event_handlers: Vec<RpcTransportEventHandlerShared>,
}
```

Same approach could be done for z_rpc:
```rust
struct ZRpcClient {
  active_channel_id: usize,
  channels: Vec<CompactTxStreamerClient<Channel>>,
}
```

@laruh does this make sense? Or am I being too naive?

Reply:
> So what I want to say, may be we should have a separate module for trait with Ops to get client or iterate over urls, if client doesn't work.

As I understand, you want a trait separate from tendermint for rpc clients that have 2 functions `get_rpc_client`, `iterate_over_urls` then implement this trait for both zcoin and tendermint. This actually a good idea :)

@ozkanonur do you think this is related to this issue https://github.com/KomodoPlatform/atomicDEX-API/issues/1480, If so we need to mention this required refactor in the issue and it shouldn't be done here :)

Reply:
> To be honest, I didn't quite understand what you mean, so will try to explain my opinion: Regarding #1480, I think it could be solved within a Rpc client structure, not outside of it:
> 
> ```rust
> struct Web3Transport {
>   // An identifier of the URI
>   active_rpc_id: usize,
>   id: Arc<AtomicUsize>,
>   uris: Vec<http::Uri>,
>   event_handlers: Vec<RpcTransportEventHandlerShared>,
> }
> ```
> 
> Same approach could be done for z_rpc:
> 
> ```rust
> struct ZRpcClient {
>   active_channel_id: usize,
>   channels: Vec<CompactTxStreamerClient<Channel>>,
> }
> ```
> 
> @laruh does this make sense? Or am I being too naive?

Sorry, now I see, my explanation was a mess. will provide my code examples.
So now we usually have smth like that: `Vec<CompactTxStreamerClient<Channel>>` (not a part of any structure), `Vec<Web3TransportNode>` (one of the fields from `struct Web3Transport)`, `Vec<HttpClient>` (filed from `struct TendermintCoinImpl`).
Onur suggested using only one alive client for requests instead of iteration over clients every time.
So, I suggest looking at the current implementation of the `clients vector` as 
```rust
struct TendermintRpcClient {
    rpc_urls: Vec<String>,
    rpc_client: AsyncMutex<HttpClient>,
}

struct ZRpcClient {
    lightwalletd_urls: Vec<String>,
    rpc_client: AsyncMutex<CompactTxStreamerClient<Channel>>,
}

struct Web3Node {
    urls: Vec<String>,
    node: AsyncMutex<Web3TransportNode>,
}
```
We can replace vectors of clients with `rpc structures` that contain necessary fields.
And then we could have a common trait named like `RpcCommonOps` with methods `get_rpc_client` (to get alive client) and `iterate_over_urls` (to iterate over urls if request for alive client isnt successful).
As a result, we can have a new rule in our project, when we need `rpc client` to send requests in different coin implementations: Create `structure` that contains `one client` and `vector of urls`. Implement trait `RpcCommonOps` for this structure.
In addition, we have to wrap `rpc_client` field into `Mutex` to be able to change it, if its dead.
In my example I use AsyncMutex, bcz I need the MutexGuard to be Send in tendermint implementation.

Reply:
Trait implementation for all rpc client handlings is a good idea. So we will be enforced by compiler to keep rpc client logics (at least args, function names and result types) as same as others.